llvm/lib/SYCLLowerIR/SYCLJointMatrixTransform.cpp

```diff
@@ -22,16 +22,45 @@ namespace {
 static constexpr char ACCESS_CHAIN[] = "_Z19__spirv_AccessChain";
 static constexpr char MATRIX_TYPE[] = "spirv.CooperativeMatrixKHR";
 
-// This routine extracts spirv.CooperativeMatrixKHR target extension type
-// from sycl::joint_matrix class object if it's used in __spirv_AccessChain
-// function call. It's necessary because otherwise OpAccessChain indices would
-// be wrong.
+// This function finds all calls to __spirv_AccessChain function and transforms
+// its users and operands to make LLVM IR more SPIR-V friendly.
 bool transformAccessChain(Function *F) {
   bool ModuleChanged = false;
   for (auto I : F->users()) {
     auto *CI = dyn_cast<CallInst>(I);
     if (!CI)
       continue;
+
+    // This is a W/A for bfloat16 and tf32 types - they are represented in SYCL
+    // as structures with int16/float storages. It means, that in LLVM IR
+    // user of CallInst to __spirv_AccessChain function would be not load/store
+    // instruction, but a zero GEP. This zero GEP is no-op, but can confuse a
+    // SPIR-V consumer, so lets remove it here.
+    auto *Unique = CI->getUniqueUndroppableUser();
+    if (auto *CastCand = dyn_cast_or_null<Instruction>(Unique)) {
+      if (auto *GEP = dyn_cast<GetElementPtrInst>(CastCand)) {
```

Comment:
Can it be simplified to this?
```suggestion
    if (auto *GEP = dyn_cast_or_null<GetElementPtrInst>(Unique)) {
```

Reply:
May be, I'll check that. My previous experience with llvm::User API shows, that you can't cast it to a particular instruction and has to do an intermediate cast to just Instruction, but may be I'm wrong

Reply:
I was wrong, thanks for the suggestion!